### PR TITLE
[eas-cli] move android submission command from expo-cli

### PR DESCRIPTION
**Pull request by barthap**
Currently, it is almost 1:1 with `expo upload:android`, but removed local uploads, now it uses only submission service.
Now it is `eas build:submit --platform android`.

The code will be probably refactored somehow when adding the iOS part. Uploading latest build is not implemented yet, I'll add it right after merging the grapgql endpoint on www.

### CODE REVIEW

File: packages/eas-cli/src/commands/build/submit.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
+import { Command, flags } from '@oclif/command';
+import { SubmissionPlatform } from '../../submissions/types';
+
+import AndroidSubmitCommand from '../../submissions/android/AndroidSubmitCommand';
+import { AndroidSubmitCommandFlags } from '../../submissions/android/types';
+import { findProjectRootAsync } from '../../project/projectUtils';
+
+export default class SubmitAndroid extends Command {
```
Comment (wkozyra95) [inline packages/eas-cli/src/commands/build/submit.ts:8]:
```suggestion
export default class Submit extends Command {
```

File: packages/eas-cli/src/commands/build/submit.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+      default: 'internal',
+      options: ['production', 'beta', 'alpha', 'internal', 'rollout'],
+    }),
+    'release-status': flags.enum({
+      description:
+        'Release status (used when uploading new apks/aabs), choose from: completed, draft, halted, inProgress',
+      default: 'completed',
+      options: ['completed', 'draft', 'halted', 'inProgress'],
+    }),
+
```
Comment (wkozyra95) [inline packages/eas-cli/src/commands/build/submit.ts:61]: Android only flags should be clearly marked

File: packages/eas-cli/src/standaloneBuilds.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  platform: Platform;
+  id?: string;
+  slug: string;
+  owner?: string;
+}
+
+export type Build = any;
+
+// Legacy Turtle v1 builds
+export async function getStandaloneBuilds(
```
Comment (wkozyra95) [inline packages/eas-cli/src/standaloneBuilds.ts:16]: I think that we decided not to support old builds?

File: packages/eas-cli/src/standaloneBuilds.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  { platform, slug, owner, id }: StandaloneBuildParams,
+  limit?: number
+): Promise<Build[]> {
+  await ensureLoggedInAsync();
```
Comment (wkozyra95) [inline packages/eas-cli/src/standaloneBuilds.ts:20]: the function that is essentially api call wrapper shouldn't be triggering login prompt

File: packages/eas-cli/src/submissions/utils/config.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import { ExpoConfig, getConfig } from '@expo/config';
+
+interface AppConfig {
+  owner?: string;
+  slug: string;
+}
+
+export function getAppConfig(projecDir: string): AppConfig {
+  const exp = getExpoConfig(projecDir);
+  if (!exp.slug) {
```
Comment (wkozyra95) [inline packages/eas-cli/src/submissions/utils/config.ts:10]: I think that slug can't be missing if it's not present in app.json name of the directory is used

File: packages/eas-cli/src/submissions/utils/files.ts [hunk 6]
```diff
@@ -0,0 +1,126 @@
+import { sync as globSync } from 'fast-glob';
```
Comment (wkozyra95) [inline packages/eas-cli/src/submissions/utils/files.ts:1]: fast glob supports async

File: packages/eas-cli/src/standaloneBuilds.ts (hunk 3 above)
Comment (barthap) [reply packages/eas-cli/src/standaloneBuilds.ts:16]: Leftover after rebasing, the file should be deleted ofc

File: packages/eas-cli/src/submissions/utils/config.ts (hunk 5 above)
Comment (barthap) [reply packages/eas-cli/src/submissions/utils/config.ts:10]: I didn't dig into `@expo/config` internals, I just assumed this code is correct as it was used in `expo-cli`. In theory, `exp.slug` is of type `string` (non-nullable), but a double check won't hurt.

File: packages/eas-cli/package.json [hunk 7]
```diff
@@ -34,9 +36,11 @@
     "pkg-dir": "^4.2.0",
     "progress": "^2.0.3",
     "prompts": "^2.3.2",
+    "tar": "^6.0.5",
     "tslib": "^1",
     "untildify": "^4.0.0",
     "uuid": "^8.3.0",
+    "validator": "^13.1.17",
```
Comment (dsokal) [inline packages/eas-cli/package.json:45]: As far as I can see, we're only using it for UUID validation. 
Let's use https://github.com/uuidjs/uuid#uuidvalidatestr instead.
```suggestion
```

File: packages/eas-cli/package.json [hunk 8]
```diff
@@ -54,7 +58,9 @@
     "@types/node-forge": "^0.9.5",
     "@types/progress": "^2.0.3",
     "@types/prompts": "^2.0.8",
+    "@types/tar": "^4.0.3",
     "@types/uuid": "^8.3.0",
+    "@types/validator": "^13.1.0",
```
Comment (dsokal) [inline packages/eas-cli/package.json:65]:
```suggestion
```

File: packages/eas-cli/src/commands/build/submit.ts [hunk 9]
```diff
@@ -0,0 +1,102 @@
+import { Command, flags } from '@oclif/command';
+import { SubmissionPlatform } from '../../submissions/types';
+
+import AndroidSubmitCommand from '../../submissions/android/AndroidSubmitCommand';
+import { AndroidSubmitCommandFlags } from '../../submissions/android/types';
+import { findProjectRootAsync } from '../../project/projectUtils';
+
+export default class BuildSubmit extends Command {
+  static description = 'Submits built artifact to app store';
```
Comment (dsokal) [inline packages/eas-cli/src/commands/build/submit.ts:9]:
```suggestion
  static description = 'Submits build artifact to app store';
```

File: packages/eas-cli/src/submissions/SubmissionService.ts [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    .json();
+  return data.submissionId;
+}
+
+async function getSubmissionAsync(projectId: string, submissionId: string): Promise<Submission> {
+  const { data } = await apiClient
+    .get(`projects/${projectId}/app-store-submissions/${submissionId}`)
+    .json();
+  return data;
+}
```
Comment (dsokal) [inline packages/eas-cli/src/submissions/SubmissionService.ts:35]: It'd be great to move to GraphQL in follow-up PRs.

File: packages/eas-cli/src/submissions/android/AndroidSubmissionConfig.ts [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+export enum AndroidArchiveType {
+  apk = 'apk',
+  aab = 'aab',
+}
+export enum IosArchiveType {
+  ipa = 'ipa',
+}
+
+export type ArchiveType = AndroidArchiveType | IosArchiveType;
```
Comment (dsokal) [inline packages/eas-cli/src/submissions/android/AndroidSubmissionConfig.ts:19]: This shouldn't be inside `android` directory.

Comment (dsokal) [reply packages/eas-cli/src/submissions/android/AndroidSubmissionConfig.ts:19]: Also, this seems irrelevant to moving the Android part of `upload:*`. I feel it shouldn't be included in this PR.

File: packages/eas-cli/src/submissions/android/AndroidSubmitCommand.ts [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+      androidPackageSource: androidPackageSource.enforceValue(),
+      track: track.enforceValue(),
+      releaseStatus: releaseStatus.enforceValue(),
+      archiveSource: archiveSource.enforceValue(),
+      serviceAccountSource: serviceAccountSource.enforceValue(),
+    };
+  }
+
+  private async getProjectIdAsync(): Promise<string> {
+    const user = await ensureLoggedInAsync();
```
Comment (dsokal) [inline packages/eas-cli/src/submissions/android/AndroidSubmitCommand.ts:75]: Maybe pre-authenticate the user in the `run` function - https://github.com/expo/eas-cli/pull/30/files#diff-f916f1e07c79cf8c4c17df1404389994227369f20d0f3ce47280aef8b5cf4965R70 - to be sure that if the user is not logged in he cannot use this command.

File: packages/eas-cli/src/submissions/android/AndroidSubmitCommand.ts [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+    const exp = getExpoConfig(this.ctx.projectDir);
+    return await ensureProjectExistsAsync({
+      accountName: exp.owner || user.username,
```
Comment (dsokal) [inline packages/eas-cli/src/submissions/android/AndroidSubmitCommand.ts:78]: Use `getProjectAccountNameAsync` from `src/project/projectUtils` instead.

File: packages/eas-cli/src/submissions/android/AndroidSubmitter.ts [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+  archive: Archive;
+  serviceAccountPath: string;
+}
+
+class AndroidSubmitter {
+  constructor(private ctx: AndroidSubmissionContext, private options: AndroidSubmissionOptions) {}
+
+  async submitAsync(): Promise<void> {
+    const resolvedSourceOptions = await this.resolveSourceOptions();
+
```
Comment (dsokal) [inline packages/eas-cli/src/submissions/android/AndroidSubmitter.ts:42]:
```suggestion
```

File: packages/eas-cli/src/submissions/android/__tests__/AndroidSubmitCommand-test.ts [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+
+jest.mock('fs');
+jest.mock('../../SubmissionService');
+jest.mock('../../../project/ensureProjectExists');
+jest.mock('@expo/image-utils', () => ({
+  generateImageAsync(input: any, { src }: any) {
+    const fs = require('fs');
+    return { source: fs.readFileSync(src) };
+  },
+}));
```
Comment (dsokal) [inline packages/eas-cli/src/submissions/android/__tests__/AndroidSubmitCommand-test.ts:28]: I think we don't need it here :)

File: packages/eas-cli/src/submissions/android/types.ts [hunk 16]
```diff
@@ -0,0 +1,11 @@
...
+
+export interface AndroidSubmitCommandFlags extends SubmitCommandFlags {
+  type?: 'apk' | 'aab';
+  key?: string;
+  androidPackage?: string;
+  track: string;
+  releaseStatus: string;
+}
+
+export type AndroidSubmissionContext = SubmissionContext<AndroidSubmitCommandFlags>;
```
Comment (dsokal) [inline packages/eas-cli/src/submissions/android/types.ts:11]: Do we still have to have these types in a separate file?

File: packages/eas-cli/src/submissions/archive-source/ArchiveTypeSource.ts [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+
+export async function getArchiveTypeAsync(
+  platform: SubmissionPlatform,
+  source: ArchiveTypeSource,
+  location: string
+): Promise<ArchiveType> {
+  // for iOS we have only one archive type
+  if (platform === SubmissionPlatform.iOS) {
+    return IosArchiveType.ipa;
+  }
```
Comment (dsokal) [inline packages/eas-cli/src/submissions/archive-source/ArchiveTypeSource.ts:46]: See my other comment - this seems irrelevant to the Android part.

File: packages/eas-cli/src/submissions/utils/builds.ts [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+  }
+
+  const {
+    platform: buildPlatform,
+    artifacts: { buildUrl },
+  } = data.builds.byId;
+
+  if (buildPlatform !== graphqlPlatform[platform]) {
+    throw new Error("Build platform doesn't match!");
+  }
```
Comment (dsokal) [inline packages/eas-cli/src/submissions/utils/builds.ts:42]:
```suggestion
```
I think we don't want to check that. If something like this happened it would mean that our API is broken.

File: packages/eas-cli/src/submissions/utils/builds.ts [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+
+const graphqlPlatform: Record<SubmissionPlatform, string> = {
+  [SubmissionPlatform.Android]: 'ANDROID',
+  [SubmissionPlatform.iOS]: 'IOS',
+};
+
+export async function getBuildArtifactUrlByIdAsync(
+  platform: SubmissionPlatform,
+  buildId: string
+): Promise<string> {
```
Comment (dsokal) [inline packages/eas-cli/src/submissions/utils/builds.ts:14]:
```suggestion
): Promise<Build> {
```
Maybe this function should return the build identified by an id? This would make the function reusable in other places.

File: packages/eas-cli/src/submissions/utils/builds.ts [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+    throw new Error("Build platform doesn't match!");
+  }
+
+  return buildUrl;
+}
+
+export async function getLatestBuildArtifactUrlAsync(
+  platform: SubmissionPlatform,
+  appId: string
+): Promise<string> {
```
Comment (dsokal) [inline packages/eas-cli/src/submissions/utils/builds.ts:50]:
```suggestion
): Promise<Build> {
```
Same here.

File: packages/eas-cli/src/submissions/utils/builds.ts [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+    )
+    .toPromise();
+
+  if (error?.graphQLErrors) {
+    throw error.graphQLErrors[0];
+  }
+
+  if (data.builds.allForApp.length !== 1) {
+    throw new Error(`No builds were found for platform: ${platform}`);
+  }
```
Comment (dsokal) [inline packages/eas-cli/src/submissions/utils/builds.ts:77]: Should this really throw an error? Maybe `null` or `undefined` would be better.

File: packages/eas-cli/src/submissions/utils/config.ts [hunk 22]
```diff
@@ -0,0 +1,26 @@
...
+    throw new Error('expo.slug is not defined in app.json');
+  }
+  return {
+    owner: exp.owner,
```
Comment (dsokal) [inline packages/eas-cli/src/submissions/utils/config.ts:14]:
```suggestion
    owner: exp.owner,
```
See the other comment about using `getProjectAccountNameAsync`

File: packages/eas-cli/src/submissions/utils/config.ts [hunk 23]
```diff
@@ -0,0 +1,26 @@
...
+}
+
+let exp: ExpoConfig;
+export function getExpoConfig(projectDir: string): ExpoConfig {
+  if (!exp) {
+    const { exp: _exp } = getConfig(projectDir, { skipSDKVersionRequirement: true });
+    exp = _exp;
+  }
+  return exp;
+}
```
Comment (dsokal) [inline packages/eas-cli/src/submissions/utils/config.ts:26]: If this function is needed it should live in some global utils directory.

**Review by dsokal — CHANGES_REQUESTED** [review]

File: packages/eas-cli/src/submissions/SubmissionService.ts (hunk 10 above)
Comment (barthap) [reply packages/eas-cli/src/submissions/SubmissionService.ts:35]: That's what I'm going to do soon.

File: packages/eas-cli/src/submissions/utils/builds.ts (hunk 19 above)
Comment (barthap) [reply packages/eas-cli/src/submissions/utils/builds.ts:14]: I'm going to move and refactor those functions when building new GraphQL foundation that we discussed.

File: packages/eas-cli/src/submissions/utils/builds.ts (hunk 18 above)
Comment (barthap) [reply packages/eas-cli/src/submissions/utils/builds.ts:42]: We have to, because we cannot filter it in gql `builds.byId` query. This is called when user specifies build ID on his own - he may copy invalid ID by mistake.

File: packages/eas-cli/package.json (hunk 7 above)
Comment (barthap) [reply packages/eas-cli/package.json:45]: We also use it for validating URL: https://github.com/expo/eas-cli/blob/a28ffba4580b2e87b8c720ca124fc154980c08b2/packages/eas-cli/src/submissions/archive-source/ArchiveFileSource.ts#L264

Comment (dsokal) [reply packages/eas-cli/package.json:45]: Hmm, maybe let's find some way that doesn't use `validator.isURL` for URL validation. Maybe https://nodejs.org/api/url.html can help us.

File: packages/eas-cli/src/submissions/utils/builds.ts (hunk 18 above)
Comment (dsokal) [reply packages/eas-cli/src/submissions/utils/builds.ts:42]: Ah, I see

File: packages/eas-cli/src/submissions/archive-source/ArchiveFileSource.ts [hunk 24]
```diff
@@ -0,0 +1,284 @@
+import chalk from 'chalk';
+import { URL, parse as parseUrl } from 'url';
+import { validate as uuidValidate } from 'uuid';
```
Comment (dsokal) [inline packages/eas-cli/src/submissions/archive-source/ArchiveFileSource.ts:3]:
```suggestion
import { validate as validateUuid } from 'uuid';
```
or 
```suggestion
import * as uuid from 'uuid';
```
or just 
```suggestion
import uuid from 'uuid';
```
if that works.

**Review by dsokal — APPROVED** [review]
Just one comment inline.
Other than that looks good. Great job!

**Review by wkozyra95 — APPROVED** [review]
